### ISSUE TITLE
Custom search path for afanasy Services, Parsers and Resources scripts?

### ISSUE THREAD

**Issue by hradec** [CGRU/cgru#248]
We're integrating afanasy as the main renderfarm backend for our studio pipeline called PipeVFX.

As one of the main features in our pipe is an  hierarquical version control system  of libraries and apps for jobs, we're doing the afanasy integration in the very same way, specially regarding custom integration code, like parsers and resoures.

We would very much like to keep afanasy code untouched in its own versioned folder so upgrading to a new version is simple. 

For that to work, we need to put the custom parser and resource scripts in a separated place, but we don't known how to make afanasy load those.

Is there some envinronment variable like AF_PARSER_PATH or somehing similar so we can add our custom path to afanasy search path?

Basically a functionality similar to MAYA_MEL_SCRIPT env var, were maya sources all paths in the search path.

Make sense?

**Comment by eliemichel** [comment]
I think you can just add the directory containing `parser/` and `services/` to the `PYTHONPATH` env var. Did you try this?

**Comment by timurhai** [comment]
Hi.
Also you can just put a new class to an existing folders.
And if you want to change an existing class, and if such improvement is not "in-house", you can just make a pull request and share new features with others.

**Comment by hradec** [comment]
yeah... for sure we're going to do some pull requests soon... but there is some overlapping in our current pipeline with some of the functionality in your parsers, so we are probably going to write some 'bridge' parser just to "glue" or frame-checking/error detection code with afanasy.

But the more we integrate afanasy with PipeVFX, the more we will be able to modify our code to make snippets that can be used by booth projects.

FYI, I have being thinking that Afanasy/Rules are kinda complemetary to PipeVFX(here on github). In PipeVFX we have focused tons of effort in job structure with app version control. We also have a rudimentary asset mechanism being build as well, and we also focus on an app agnostic methodology of work, using CortexVFX and Gaffer (booth here in github) to pass data between Maya, Houdini, Nuke, etc.

Afanasy is the perfect farm mechanism for us, and we're seriously considering Rules as not only a production manager, but also as a front end for our asset management.

So, I'll be asking some more questions in the future for sure!

Regarding pythonpath, I haven't tried yet... I'll do for sure! Thanks!

**Comment by hradec** [comment]
so, by adding my custom search path to PYTHONPATH, Do I have to create a "parser" folder module in the same way as the one in afanasy/python path?

if so, wouldn't that REPLACE the original one?

**Comment by timurhai** [comment]
Yes.
Yes.
( this is a common Python behavior, nothing Afanasy special )

**Comment by hradec** [comment]
well, thats not exactly what we want to do. 

we want to just ADD functionality by searchpath, not replace it! 

We can add a searchpath functionality to parsers, services and resources and pull request it then.

Would AF_PARSER_PATH, AF_RESOURCE_PATH and AF_SERVICES_PATH make sense as search path env vars for you?

**Comment by timurhai** [comment]
So you chould ADD a custom path to PYTHONPATH, not replace it! :)

**Comment by timurhai** [comment]
You can look at Python docs, especially about PYTHONPATH variable.
How Python search and load modules is not a CGRU issue.

**Comment by hradec** [comment]
I did add a custom path to pythonpath, but if my custom path is first, when Python import the parser module, for example, it will import the one in my custom path and stop, ignoring the default one in afanasy. (thats how python deals with pythonpath)

So to have all the parsers from cgru + my customs ones, I would have to copy all parsers from cgru to my parser module in my custom path, which defeats the purpose! 

make sense? Im I missing something?

**Comment by eliemichel** [comment]
You say that Python imports the parser which is in your directory. This means that you have a file called `parsers/parser.py`. What is this module doing? Is it an afanasy parser or something completely unrelated that happens to shadow afanasy's because you use a similar design for something different?

First case, it is an afanasy parser. In that case, why calling it just `parser.py` if it is _not_ supposed to shadow/replace afanasy's defautl parser? If it is not a default parser, just give it a more relevant name. As any other module in `parsers/` does.

Second case, it is something unrelated. In that case I think you should be able to manage to set differently the `PYTHONPATH` in your different units of code. The code calling afanasy adds afanasy's path. The code using the other unrelated features sets its `PYTHONPATH` differently. In the worst case, you'll just have to play with `sys.path`.

Adding yet another set of env variables would just overcomplicate the design imho. The use of env vars should be minimized. Having similar information provided by different vars would just make it possible to have weird inconsistent states.

**Comment by hradec** [comment]
Ok... after examining the code I finally figure why I was having trouble to understand what you're saying! 

in file afanasy/python/check_parser.py, it uses a __import__ call to import the required parser for a given job, were it looks for it INSIDE a "parsers" folder. As you're using the __import__  call, it will use the PYTHONPATH to search from a parsers/<file>!!  
`
 __import__('parsers', globals(), locals(), [parserType])
`

I was having trouble to understand how afanasy could do that with the standard python "import", but as you're using __import__, it works!!

In this case, I totally agree... PYTHONPATH has to work!! 

Sorry my mis-understading... Really appreciate the chat though!! :)

cheers...
